File: src/core/gameData.ts
import { parseCsv, tryParseInt, type CsvFiles } from './csv';

export const Difficulty = { Easy: 'easy', Medium: 'medium', Hard: 'hard' } as const;
export type Difficulty = (typeof Difficulty)[keyof typeof Difficulty];

export const Channel = { Environment: 'environment', Phone: 'phone' } as const;
export type Channel = (typeof Channel)[keyof typeof Channel];

/** sources.csv 的一行（4.2 环境答案源清单的数据化）。 */
export interface AnswerSource {
  id: string;
  name: string;
  channelLabel: string;
  location: string;
  riskLabel: string;
  demo: string;
  notes: string;
  line: number;
  occlusions: string[];
  channel: Channel;
  riskWeight: number;
  isDemo: boolean;
}

/** answer_chain.csv 的一行（4.3 的关卡数据）。 */
export interface AnswerChainRow {
  qId: string;
  answerTypeLabel: string;
  answer: string;
  /** 动态题的机器可读答案（AnswerExpression 语法）；固定题为空，答案就是 answer。 */
  answerExpr: string;
  /** 四选一的干扰项（决策 #38）：字面列表 / auto:偏移 / rest。 */
  distractors: string;
  crossSource: string;
  difficultyLabel: string;
  demo: string;
  notes: string;
  line: number;
  pools: string[];
  occlusions: string[];
  difficulty: Difficulty;
  isDemo: boolean;
  isDynamic: boolean;
}

/** occlusion_ids.csv 的一行：把中文遮挡状态映射成代码可用的 ASCII 枚举名。 */
export interface OcclusionDef {
  label: string;
  id: string;
  kind: string;
  extraSteps: number;
  /** false ⇒ 信息不可得，必须换到该题另一个落位池。 */
  resolvable: boolean;
  note: string;
}

/** enum_labels.csv：中文标签 → ASCII id / 权重。让源码里可以一个中文字面量都没有。 */
export class EnumLabels {
  private readonly ids = new Map<string, string>();
  private readonly weights = new Map<string, number>();
  readonly problems: string[] = [];

  add(domain: string, label: string, id: string, weight: number): void {
    this.ids.set(domain + '|' + label, id);
    this.weights.set(domain + '|' + label, weight);
  }

  id(domain: string, label: string): string {
    const v = this.ids.get(domain + '|' + label);
    if (v !== undefined) return v;
    this.problems.push('enum_labels.csv missing ' + domain + ' -> label(line ' + label.length + ' chars)');
    return 'unknown';
  }

  weight(domain: string, label: string): number {
    const v = this.weights.get(domain + '|' + label);
    if (v !== undefined) return v;
    this.problems.push('enum_labels.csv missing weight for domain ' + domain);
    return 0;
  }

  toDifficulty(label: string): Difficulty {
    const id = this.id('difficulty', label);
    if (id === 'easy') return Difficulty.Easy;
    if (id === 'hard') return Difficulty.Hard;
    return Difficulty.Medium;
  }

  toChannel(label: string): Channel {
    return this.id('channel', label) === 'phone' ? Channel.Phone : Channel.Environment;
  }
}

export const SOURCE_HEADER = ['id', 'name', 'channel', 'location', 'risk', 'demo', 'occlusions', 'notes'];
export const CHAIN_HEADER = ['q_id', 'answer_type', 'answer', 'answer_expr', 'distractors', 'pools', 'cross_source', 'occlusions', 'difficulty', 'demo', 'notes'];
export const OCCLUSION_HEADER = ['occlusion', 'id', 'kind', 'extra_steps', 'resolvable', 'note'];
export const ENUM_LABEL_HEADER = ['domain', 'label', 'id', 'weight', 'note'];

export class GameData {
  readonly sources: AnswerSource[] = [];
  readonly sourceById = new Map<string, AnswerSource>();
  readonly chain: AnswerChainRow[] = [];
  readonly occlusionByLabel = new Map<string, OcclusionDef>();
  readonly labels = new EnumLabels();
  readonly problems: string[] = [];

  /** files 以文件名为 key（例 "sources.csv"），即 Assets/Data/ 下的全部 CSV。 */
  static load(files: CsvFiles): GameData {
    const d = new GameData();
    d.loadEnumLabels(files);
    d.loadOcclusions(files);
    d.loadSources(files);
    d.loadChain(files);
    d.crossCheck();
    return d;
  }

  private read(files: CsvFiles, name: string, header: string[]): string[][] | null {
    const text = files[name];
    if (text === undefined) { this.problems.push('missing file: ' + name); return null; }
    const rows = parseCsv(text);
    if (rows.length === 0) { this.problems.push(name + ': empty'); return null; }
    if (rows[0].join(',') !== header.join(',')) { this.problems.push(name + ': header mismatch'); return null; }
    return rows;
  }

  private loadEnumLabels(files: CsvFiles): void {
    const rows = this.read(files, 'enum_labels.csv', ENUM_LABEL_HEADER);
    if (!rows) return;
    for (let i = 1; i < rows.length; i++) {
      const r = rows[i];
      if (r.length !== ENUM_LABEL_HEADER.length) { this.problems.push('enum_labels.csv:' + (i + 1) + ': bad column count'); continue; }
      this.labels.add(r[0], r[1], r[2], tryParseInt(r[3]) ?? 0);
    }
  }

  private loadOcclusions(files: CsvFiles): void {
    const rows = this.read(files, 'occlusion_ids.csv', OCCLUSION_HEADER);
    if (!rows) return;
    for (let i = 1; i < rows.length; i++) {
      const r = rows[i];
      if (r.length !== OCCLUSION_HEADER.length) { this.problems.push('occlusion_ids.csv:' + (i + 1) + ': bad column count'); continue; }
      const def: OcclusionDef = {
        label: r[0],
        id: r[1],
        kind: r[2],
        extraSteps: tryParseInt(r[3]) ?? 0,
        resolvable: r[4] !== 'n',
        note: r[5],
      };
      if (this.occlusionByLabel.has(def.label)) { this.problems.push('occlusion_ids.csv: duplicate label at line ' + (i + 1)); continue; }
      this.occlusionByLabel.set(def.label, def);
    }
  }

  private loadSources(files: CsvFiles): void {
    const rows = this.read(files, 'sources.csv', SOURCE_HEADER);
    if (!rows) return;
    for (let i = 1; i < rows.length; i++) {
      const r = rows[i];
      if (r.length !== SOURCE_HEADER.length) { this.problems.push('sources.csv:' + (i + 1) + ': bad column count'); continue; }
      const s: AnswerSource = {
        id: r[0],
        name: r[1],
        channelLabel: r[2],
        location: r[3],
        riskLabel: r[4],
        demo: r[5],
        notes: r[7],
        line: i + 1,
        occlusions: splitPipes(r[6]),
        channel: this.labels.toChannel(r[2]),
        riskWeight: this.labels.weight('risk', r[4]),
        isDemo: r[5] === 'y',
      };
      if (this.sourceById.has(s.id)) { this.problems.push('sources.csv: duplicate id ' + s.id); continue; }
      this.sourceById.set(s.id, s);
      this.sources.push(s);
    }
  }

  private loadChain(files: CsvFiles): void {
    const rows = this.read(files, 'answer_chain.csv', CHAIN_HEADER);
    if (!rows) return;
    for (let i = 1; i < rows.length; i++) {
      const r = rows[i];
      if (r.length !== CHAIN_HEADER.length) { this.problems.push('answer_chain.csv:' + (i + 1) + ': bad column count'); continue; }
      const answerType = this.labels.id('answer_type', r[1]);
      this.chain.push({
        qId: r[0],
        answerTypeLabel: answerType,
        answer: r[2],
        answerExpr: r[3],
        distractors: r[4],
        pools: splitPipes(r[5]),
        crossSource: r[6],
        occlusions: splitPipes(r[7]),
        difficultyLabel: r[8],
        demo: r[9],
        notes: r[10],
        line: i + 1,
        difficulty: this.labels.toDifficulty(r[8]),
        isDemo: r[9] === 'y',
        isDynamic: answerType === 'dynamic',
      });
    }
  }

  /** 参照完整性检查：JS 校验器查不到的、跨文件的东西在这里查。 */
  private crossCheck(): void {
    for (const s of this.sources) {
      for (const occ of s.occlusions) {
        if (!this.occlusionByLabel.has(occ)) this.problems.push('sources.csv:' + s.line + ': occlusion not in occlusion_ids.csv -> ' + occ);
      }
    }
    for (const c of this.chain) {
      for (const p of c.pools) {
        if (!this.sourceById.has(p)) this.problems.push('answer_chain.csv:' + c.line + ': unknown pool source -> ' + p);
      }
      for (const occ of c.occlusions) {
        if (!this.occlusionByLabel.has(occ)) this.problems.push('answer_chain.csv:' + c.line + ': occlusion not in occlusion_ids.csv -> ' + occ);
      }
      // Demo 落位可达性：demo=y 的题，落位池里至少要有一个 demo=y 的源，否则它在 Demo 里没有合法落位。
      const demoPools = this.poolSources(c).filter((s) => s.isDemo).length;
      if (c.isDemo && c.pools.length > 0 && demoPools === 0) {
        this.problems.push('answer_chain.csv:' + c.line + ': demo=y but every pool source is demo=n -> no legal placement in the demo: ' + c.qId);
      }
      // cross_source 必须是「源 id」（可带括号说明），否则代码没法解析它。
      if (c.crossSource.length > 0) {
        for (const x of c.crossSource.split('|')) {
          const id = stripDescription(x);
          if (id.length > 0 && !this.sourceById.has(id)) this.problems.push('answer_chain.csv:' + c.line + ': cross_source is not a known source id -> ' + x);
        }
      }
    }
  }

  poolSources(c: AnswerChainRow): AnswerSource[] {
    const list: AnswerSource[] = [];
    for (const p of c.pools) {
      const s = this.sourceById.get(p);
      if (s) list.push(s);
    }
    return list;
  }

  /** 该题的落位池是否跨了两条通道（4.2：两条通道必须互相不可替代）。 */
  isMixedChannel(c: AnswerChainRow): boolean {
    let env = false;
    let phone = false;
    for (const s of this.poolSources(c)) {
      if (s.channel === Channel.Phone) phone = true;
      else env = true;
    }
    return env && phone;
  }

  countByDifficulty(d: Difficulty, demoOnly: boolean): number {
    return this.chain.filter((c) => (!demoOnly || c.isDemo) && c.difficulty === d).length;
  }
}

function splitPipes(raw: string): string[] {
  if (!raw) return [];
  return raw.split('|').map((p) => p.trim()).filter((p) => p.length > 0);
}

const FULLWIDTH_PAREN = String.fromCharCode(0xff08);

/** 取「源 id」本体：允许 "eraser（借橡皮金属反光...）" 这种带说明的写法。 */
function stripDescription(raw: string): string {
  let s = (raw ?? '').trim();
  let cut = s.indexOf('(');
  const full = s.indexOf(FULLWIDTH_PAREN);
  if (cut < 0 || (full >= 0 && full < cut)) cut = full;
  if (cut >= 0) s = s.substring(0, cut);
  return s.trim();
}

File: Assets/Scripts/APlus/GameData.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace APlus
{
    public enum Difficulty { Easy, Medium, Hard }
    public enum Channel { Environment, Phone }

    /// <summary>sources.csv 的一行（4.2 环境答案源清单的数据化）。</summary>
    public sealed class AnswerSource
    {
        public string Id = "";
        public string Name = "";
        public string ChannelLabel = "";
        public string Location = "";
        public string RiskLabel = "";
        public string Demo = "";
        public string Notes = "";
        public int Line;
        public List<string> Occlusions = new List<string>();

        public Channel Channel;
        public int RiskWeight;
        public bool IsDemo { get { return Demo == "y"; } }
    }

    /// <summary>answer_chain.csv 的一行（4.3 的关卡数据）。</summary>
    public sealed class AnswerChainRow
    {
        public string QId = "";
        public string AnswerTypeLabel = "";
        public string Answer = "";
        /// <summary>动态题的机器可读答案（AnswerExpression 语法）；固定题为空，答案就是 Answer。</summary>
        public string AnswerExpr = "";
        /// <summary>四选一的干扰项（决策 #38）：字面列表 / auto:偏移 / rest。</summary>
        public string Distractors = "";
        public string CrossSource = "";
        public string DifficultyLabel = "";
        public string Demo = "";
        public string Notes = "";
        public int Line;
        public List<string> Pools = new List<string>();
        public List<string> Occlusions = new List<string>();

        public Difficulty Difficulty;
        public bool IsDemo { get { return Demo == "y"; } }
        public bool IsDynamic { get { return AnswerTypeLabel == "dynamic"; } }
    }

    /// <summary>occlusion_ids.csv 的一行：把 sources.csv 里的中文遮挡状态映射成代码可用的 ASCII 枚举名。</summary>
    public sealed class OcclusionDef
    {
        public string Label = "";     // 中文状态名（来自 sources.csv / answer_chain.csv）
        public string Id = "";        // ASCII 枚举名，代码与 JSON 里只用这个
        public string Kind = "";      // visual | audio
        public int ExtraSteps;        // 该遮挡额外要求的玩家动作数（原型模型）
        public bool Resolvable;       // false ⇒ 信息不可得，必须换到该题另一个落位池
        public string Note = "";
    }

    /// <summary>enum_labels.csv：中文标签 → ASCII id / 权重。让 C# 源码里可以一个中文字面量都没有。</summary>
    public sealed class EnumLabels
    {
        readonly Dictionary<string, string> _ids = new Dictionary<string, string>(StringComparer.Ordinal);
        readonly Dictionary<string, int> _weights = new Dictionary<string, int>(StringComparer.Ordinal);
        public List<string> Problems = new List<string>();

        public void Add(string domain, string label, string id, int weight)
        {
            _ids[domain + "|" + label] = id;
            _weights[domain + "|" + label] = weight;
        }

        public string Id(string domain, string label)
        {
            string v;
            if (_ids.TryGetValue(domain + "|" + label, out v)) return v;
            Problems.Add("enum_labels.csv missing " + domain + " -> label(line " + label.Length + " chars)");
            return "unknown";
        }

        public int Weight(string domain, string label)
        {
            int v;
            if (_weights.TryGetValue(domain + "|" + label, out v)) return v;
            Problems.Add("enum_labels.csv missing weight for domain " + domain);
            return 0;
        }

        public Difficulty ToDifficulty(string label)
        {
            string id = Id("difficulty", label);
            if (id == "easy") return Difficulty.Easy;
            if (id == "hard") return Difficulty.Hard;
            return Difficulty.Medium;
        }

        public Channel ToChannel(string label)
        {
            return Id("channel", label) == "phone" ? Channel.Phone : Channel.Environment;
        }
    }

    public sealed class GameData
    {
        public static readonly string[] SourceHeader = new string[]
        { "id", "name", "channel", "location", "risk", "demo", "occlusions", "notes" };

        public static readonly string[] ChainHeader = new string[]
        { "q_id", "answer_type", "answer", "answer_expr", "distractors", "pools", "cross_source", "occlusions", "difficulty", "demo", "notes" };

        public static readonly string[] OcclusionHeader = new string[]
        { "occlusion", "id", "kind", "extra_steps", "resolvable", "note" };

        public static readonly string[] EnumLabelHeader = new string[]
        { "domain", "label", "id", "weight", "note" };

        public List<AnswerSource> Sources = new List<AnswerSource>();
        public Dictionary<string, AnswerSource> SourceById = new Dictionary<string, AnswerSource>(StringComparer.Ordinal);
        public List<AnswerChainRow> Chain = new List<AnswerChainRow>();
        public Dictionary<string, OcclusionDef> OcclusionByLabel = new Dictionary<string, OcclusionDef>(StringComparer.Ordinal);
        public EnumLabels Labels = new EnumLabels();
        public List<string> Problems = new List<string>();
        public string DataDir = "";

        public static GameData Load(string dataDir)
        {
            GameData d = new GameData();
            d.DataDir = dataDir;
            d.LoadEnumLabels(Path.Combine(dataDir, "enum_labels.csv"));
            d.LoadOcclusions(Path.Combine(dataDir, "occlusion_ids.csv"));
            d.LoadSources(Path.Combine(dataDir, "sources.csv"));
            d.LoadChain(Path.Combine(dataDir, "answer_chain.csv"));
            d.CrossCheck();
            return d;
        }

        static List<string[]> ReadOrProblem(string path, string[] header, List<string> problems)
        {
            if (!File.Exists(path)) { problems.Add("missing file: " + Path.GetFileName(path)); return null; }
            List<string[]> rows = Csv.ReadFile(path);
            if (rows.Count == 0) { problems.Add(Path.GetFileName(path) + ": empty"); return null; }
            if (string.Join(",", rows[0]) != string.Join(",", header))
            {
                problems.Add(Path.GetFileName(path) + ": header mismatch");
                return null;
            }
            return rows;
        }

        void LoadEnumLabels(string path)
        {
            List<string[]> rows = ReadOrProblem(path, EnumLabelHeader, Problems);
            if (rows == null) return;
            for (int i = 1; i < rows.Count; i++)
            {
                if (rows[i].Length != EnumLabelHeader.Length) { Problems.Add("enum_labels.csv:" + (i + 1) + ": bad column count"); continue; }
                int w;
                if (!int.TryParse(rows[i][3], out w)) w = 0;
                Labels.Add(rows[i][0], rows[i][1], rows[i][2], w);
            }
        }

        void LoadOcclusions(string path)
        {
            List<string[]> rows = ReadOrProblem(path, OcclusionHeader, Problems);
            if (rows == null) return;
            for (int i = 1; i < rows.Count; i++)
            {
                if (rows[i].Length != OcclusionHeader.Length) { Problems.Add("occlusion_ids.csv:" + (i + 1) + ": bad column count"); continue; }
                OcclusionDef def = new OcclusionDef();
                def.Label = rows[i][0];
                def.Id = rows[i][1];
                def.Kind = rows[i][2];
                int steps;
                if (!int.TryParse(rows[i][3], out steps)) steps = 0;
                def.ExtraSteps = steps;
                def.Resolvable = rows[i][4] != "n";
                def.Note = rows[i][5];
                if (OcclusionByLabel.ContainsKey(def.Label)) { Problems.Add("occlusion_ids.csv: duplicate label at line " + (i + 1)); continue; }
                OcclusionByLabel.Add(def.Label, def);
            }
        }

        void LoadSources(string path)
        {
            List<string[]> rows = ReadOrProblem(path, SourceHeader, Problems);
            if (rows == null) return;
            for (int i = 1; i < rows.Count; i++)
            {
                if (rows[i].Length != SourceHeader.Length) { Problems.Add("sources.csv:" + (i + 1) + ": bad column count"); continue; }
                AnswerSource s = new AnswerSource();
                s.Id = rows[i][0];
                s.Name = rows[i][1];
                s.ChannelLabel = rows[i][2];
                s.Location = rows[i][3];
                s.RiskLabel = rows[i][4];
                s.Demo = rows[i][5];
                s.Notes = rows[i][7];
                s.Line = i + 1;
                SplitPipes(rows[i][6], s.Occlusions);
                s.Channel = Labels.ToChannel(s.ChannelLabel);
                s.RiskWeight = Labels.Weight("risk", s.RiskLabel);
                if (SourceById.ContainsKey(s.Id)) { Problems.Add("sources.csv: duplicate id " + s.Id); continue; }
                SourceById.Add(s.Id, s);
                Sources.Add(s);
            }
        }

        void LoadChain(string path)
        {
            List<string[]> rows = ReadOrProblem(path, ChainHeader, Problems);
            if (rows == null) return;
            for (int i = 1; i < rows.Count; i++)
            {
                if (rows[i].Length != ChainHeader.Length) { Problems.Add("answer_chain.csv:" + (i + 1) + ": bad column count"); continue; }
                AnswerChainRow c = new AnswerChainRow();
                c.QId = rows[i][0];
                c.AnswerTypeLabel = Labels.Id("answer_type", rows[i][1]);
                c.Answer = rows[i][2];
                c.AnswerExpr = rows[i][3];
                c.Distractors = rows[i][4];
                SplitPipes(rows[i][5], c.Pools);
                c.CrossSource = rows[i][6];
                SplitPipes(rows[i][7], c.Occlusions);
                c.DifficultyLabel = rows[i][8];
                c.Demo = rows[i][9];
                c.Notes = rows[i][10];
                c.Line = i + 1;
                c.Difficulty = Labels.ToDifficulty(c.DifficultyLabel);
                Chain.Add(c);
            }
        }

        static void SplitPipes(string raw, List<string> into)
        {
            if (raw == null || raw.Length == 0) return;
            string[] parts = raw.Split('|');
            for (int i = 0; i < parts.Length; i++)
            {
                string p = parts[i].Trim();
                if (p.Length > 0) into.Add(p);
            }
        }

        /// <summary>原型自带的参照完整性检查：JS 校验器查不到的、跨到新文件的东西在这里查。</summary>
        void CrossCheck()
        {
            for (int i = 0; i < Sources.Count; i++)
            {
                AnswerSource s = Sources[i];
                for (int j = 0; j < s.Occlusions.Count; j++)
                {
                    if (!OcclusionByLabel.ContainsKey(s.Occlusions[j]))
                        Problems.Add("sources.csv:" + s.Line + ": occlusion not in occlusion_ids.csv -> " + s.Occlusions[j]);
                }
            }
            for (int i = 0; i < Chain.Count; i++)
            {
                AnswerChainRow c = Chain[i];
                for (int j = 0; j < c.Pools.Count; j++)
                {
                    if (!SourceById.ContainsKey(c.Pools[j]))
                        Problems.Add("answer_chain.csv:" + c.Line + ": unknown pool source -> " + c.Pools[j]);
                }
                for (int j = 0; j < c.Occlusions.Count; j++)
                {
                    if (!OcclusionByLabel.ContainsKey(c.Occlusions[j]))
                        Problems.Add("answer_chain.csv:" + c.Line + ": occlusion not in occlusion_ids.csv -> " + c.Occlusions[j]);
                }
                // Demo 落位可达性（8.3 的排期口径）：demo=y 的题，落位池里至少要有一个 demo=y 的源。
                // 否则这道题在 Demo 里**没有合法落位** —— 这是原型跑数据时抓到的真实缺口。
                List<AnswerSource> poolSources = PoolSources(c);
                int demoPools = 0;
                for (int j = 0; j < poolSources.Count; j++) if (poolSources[j].IsDemo) demoPools++;
                if (c.IsDemo && c.Pools.Count > 0 && demoPools == 0)
                    Problems.Add("answer_chain.csv:" + c.Line + ": demo=y but every pool source is demo=n -> no legal placement in the demo: " + c.QId);

                // cross_source 必须是「源 id」（可带括号说明），否则代码没法解析它。
                if (c.CrossSource.Length > 0)
                {
                    string[] xs = c.CrossSource.Split('|');
                    for (int j = 0; j < xs.Length; j++)
                    {
                        string id = StripDescription(xs[j]);
                        if (id.Length > 0 && !SourceById.ContainsKey(id))
                            Problems.Add("answer_chain.csv:" + c.Line + ": cross_source is not a known source id -> " + xs[j]);
                    }
                }

                if (c.IsDynamic && c.AnswerTypeLabel != "dynamic")
                    Problems.Add("answer_chain.csv:" + c.Line + ": answer_type did not resolve");
            }
        }

        /// <summary>取「源 id」本体：允许 "eraser（借橡皮金属反光...）" 这种带说明的写法。</summary>
        static string StripDescription(string raw)
        {
            string s = raw == null ? "" : raw.Trim();
            int cut = s.IndexOf('(');
            int full = s.IndexOf((char)0xFF08);
            if (cut < 0 || (full >= 0 && full < cut)) cut = full;
            if (cut >= 0) s = s.Substring(0, cut);
            return s.Trim();
        }

        public List<AnswerSource> PoolSources(AnswerChainRow c)
        {
            List<AnswerSource> list = new List<AnswerSource>();
            for (int i = 0; i < c.Pools.Count; i++)
            {
                AnswerSource s;
                if (SourceById.TryGetValue(c.Pools[i], out s)) list.Add(s);
            }
            return list;
        }

        /// <summary>该题的落位池是否跨了两条通道（4.2：两条通道必须互相不可替代）。</summary>
        public bool IsMixedChannel(AnswerChainRow c)
        {
            bool env = false, phone = false;
            List<AnswerSource> pool = PoolSources(c);
            for (int i = 0; i < pool.Count; i++)
            {
                if (pool[i].Channel == Channel.Phone) phone = true; else env = true;
            }
            return env && phone;
        }

        public int CountByDifficulty(Difficulty d, bool demoOnly)
        {
            int n = 0;
            for (int i = 0; i < Chain.Count; i++)
            {
                if (demoOnly && !Chain[i].IsDemo) continue;
                if (Chain[i].Difficulty == d) n++;
            }
            return n;
        }
    }
}